### ISSUE TITLE
peg-grammar is not available in threads

### ISSUE THREAD

**Issue by zevv** [janet-lang/janet#1180]
This snippet:
```janet
(ev/spawn-thread       
  (peg/match '{ :main :s* } "")) 
```
results in the following error:
```
error: grammar error in :s*, unknown rule
  in peg/match [src/core/peg.c] on line 1694
  in _spawn-thread [/tmp/t.janet] on line 2, column 3
```

**Comment by sogaiu** [comment]
Yes, although not reported in an issue AFAIK, this has been noted in discussions at one or more of the gitter / matrix channels.

It seems to me that either it should be documented behavior (along with any other caveats [1] when spawning a thread) or the current behavior should be adjusted.

---

[1] I noticed that `eval` didn't give quite the same results in my testing, and I suppose it may be a related thing.

**Comment by zevv** [comment]
I must admit I have not looked deeply at the memory model for threads so I'm not sure what the original design decisions are, but I think conceptually it might make sense to have new threads inherit the parent's dynamic bindings?

**Comment by sogaiu** [comment]
For `coro`, the fiber that's created with `fiber/new` has `:yi` as part of its sigmask, and `:i` is "inherit the environment from the current fiber".

So if I do:

```janet
(ev/thread (coro (pp (peg/match {:main :s*} ""))))
```

I get:

```
@[]
nil
```

Currently `ev/spawn-thread` and `ev/do-thread` don't take any flags, but may be it would be nice if they did?  Something like:

```janet
(ev/spawn-thread :i (peg/match {:main :s*} ""))
```

I don't suppose that would break any code -- an initial keyword in a function body presumably doesn't do much.

**Comment by bakpakin** [comment]
So the `ev/thread` function, which is what does all of the heavy lifting of setting up a new thread, takes some flags to determine what state to copy from the current thread into the new thread. I suppose it would make sense to copy over dynamic bindings as an option if specified - however, copying over the entire environment table as returned by `(curenv)`, table proto types and all, is prohibitively expensive.

> I noticed that eval didn't give quite the same results in my testing, and I suppose it may be a related thing.

Yes, eval needs to have available function definitions in it's  environment, so in order to see all available functions they would need to all be copied to the new thread.

For efficiency, the default behavior is to copy nothing that isn't needed.

**Comment by bakpakin** [comment]
For completeness, as a work around I should mention that `ev/thread` can take a fiber argument as well as a function argument - while this doesn't help if you are using the built-in `ev/do-thread` and `ev/spawn-thread` macros, you could create your own subroutine to wrap a closure or function body in a fiber that captures all of the state you were interested in.

For example

```janet
(defn wrap-closure-for-thread
  "Wrap a closure in a fiber that captures the current dynamic bindings state."
  [f]
  (def env @{})
  (each k (all-dynamics) (put env k (dyn k)))
  (fiber/setenv (fiber/new f) env))
  
# Use with ev/thread
(ev/thread (wrap-closure-for-thread (fn [&] (pp (dyn *peg-grammar*)))))
```

**Comment by amano-kenji** [comment]
I think a thread that shares memory with the main thread will not have this issue?

**Comment by amano-kenji** [comment]
Why don't you inline default-peg-grammar and let people change it with a dynamic variable or a named function parameter?

**Comment by amano-kenji** [comment]
This should be opened again.

**Comment by amano-kenji** [comment]
There are ways to fix this issue and optimize janet without abandoning `(dyn :peg-grammar)` and `default-peg-grammar`, but my personal recommendation is to simplify `peg/compile` by making it referentially transparent with an inline peg grammar.

To me, the flexibility of `(dyn :peg-grammar)` and `default-peg-grammar` is not worth the trouble and all the little details you have to think through. You can kind of carry your own shared grammar separately.

Functional programming and referential transparency are a great default mode of operation.

**Comment by sogaiu** [comment]
This approach appears to not consider compatibility.

Please note the following text from a recent [interview](https://alexalejandre.com/programming/interview-with-bakpakin/):

> I try to have empathy for the user. Frankly, I have more empathy for existing users than new users. Existing users are more important than new users. There is no corporate interest, I’m not selling anything so there’s no need to grow or sell more. We just make it good for the people who care. I’m more caring about that than making it good for myself.
>
> With making things last, the idea is enabling forwards compatibility and non-breaking changes, making sure there are resources for future programmers even when we’ve all moved on. Documentation shouldn’t disappear, forum posting should at least be public if not well organized.

As I mentioned [elsewhere](https://github.com/janet-lang/janet/pull/1703#issuecomment-3831092024):

> FWIW, I've added to default-peg-grammar before. That lets you provide things to more than one grammar by only making changes in one place.

Others may have done so too.

**Comment by amano-kenji** [comment]
Another way to fix it is to respect `(dyn :peg-grammar)`, and if it is not defined, then use inline peg grammar.

You can eliminate duplication with a C macro like `#define PEG_GRAMMAR (...)`? I don't really know C.
